File: api/src/main/java/net/techcable/sonarpet/utils/compat/AncientGuava.java
package net.techcable.sonarpet.utils.compat;

import java.util.concurrent.Executor;
import java.util.stream.Collector;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import com.google.common.util.concurrent.MoreExecutors;

public class AncientGuava implements GuavaVersion {
    /* package */ AncientGuava() {}
    @Override
    public <T> Collector<T, ?, ImmutableList<T>> immutableListCollector() {
        return Collector.<T, ImmutableList.Builder<T>, ImmutableList<T>>of(
                ImmutableList::builder,
                ImmutableList.Builder::add,
                (first, second) -> first.addAll(second.build()),
                ImmutableList.Builder::build
        );
    }

    @Override
    public <T> Collector<T, ?, ImmutableSet<T>> immutableSetCollector() {
        return Collector.<T, ImmutableSet.Builder<T>, ImmutableSet<T>>of(
                ImmutableSet::builder,
                ImmutableSet.Builder::add,
                (first, second) -> first.addAll(second.build()),
                ImmutableSet.Builder::build
        );
    }

    @Override
    public Executor directExecutor() {
        return MoreExecutors.sameThreadExecutor();
    }
}

File: api/src/main/kotlin/net/techcable/sonarpet/utils/compat/AncientGuava.kt
package net.techcable.sonarpet.utils.compat

import com.google.common.collect.ImmutableList
import com.google.common.util.concurrent.MoreExecutors
import net.techcable.sonarpet.getMethodHandle
import java.lang.invoke.MethodHandle
import java.util.concurrent.Executor
import java.util.function.BiConsumer
import java.util.function.BinaryOperator
import java.util.function.Function
import java.util.function.Supplier
import java.util.stream.Collector

object AncientGuava: GuavaVersion {
    private val immutableListCollectorInstance = Collector.of(
            Supplier { ImmutableList.builder<Any>() },
            BiConsumer<ImmutableList.Builder<Any>, Any> { builder, element -> builder.add(element) },
            BinaryOperator<ImmutableList.Builder<Any>> { first, second -> first.addAll(second.build()) },
            Function<ImmutableList.Builder<Any>, ImmutableList<Any>> { it.build() }
    )
    @Suppress("UNCHECKED_CAST")
    override fun <E> immutableListCollector() = immutableListCollectorInstance as Collector<E, *, ImmutableList<E>>

    private val SAME_THREAD_EXECUTOR_METHOD: MethodHandle = MoreExecutors::class.getMethodHandle("sameThreadExecutor")
    override fun directExecutor() = SAME_THREAD_EXECUTOR_METHOD.invoke() as Executor
}